swift/extractor/infra/Path.cpp

```diff
@@ -0,0 +1,38 @@
+#include "swift/extractor/infra/Path.h"
+#include <iostream>
+#include <unistd.h>
+
+namespace codeql {
+
+static bool shouldCanonicalize() {
+  auto preserve = getenv("CODEQL_PRESERVE_SYMLINKS");
+  if (preserve && std::string(preserve) == "true") {
+    return false;
+  }
+  preserve = getenv("SEMMLE_PRESERVE_SYMLINKS");
+  if (preserve && std::string(preserve) == "true") {
+    return false;
+  }
+  return true;
+}
+
+std::filesystem::path getCodeQLPath(std::string_view path) {
+  // TODO: this needs more testing
+  // TODO: check canonicalization of names on a case insensitive filesystems
+  std::error_code ec;
+  std::filesystem::path ret = {};
+  auto canonicalize = shouldCanonicalize();
```

Comment:
should we maybe make this `static const` so reading the environment is done exactly once?

Reply:
Done.